### PR TITLE
cluster

### PR DESCRIPTION
**Pull request by scardena**
Adds cluster functionalities and integration tests for each functionality.
Adds 1 node to the tests to make stardog work as a cluster (so 2 nodes total)
Adds 1 standby node to test standby features (ad the end of the tests it will end up with 3 nodes: 2 main nodes, plus the standby)

### CODE REVIEW

File: stardog/admin.py [hunk 1]
```diff
@@ -649,68 +647,91 @@ def validate(self):
         """
         self.client.get('/admin/users/valid')
 
-    #TODO
```
Comment (buzztroll) [inline stardog/admin.py:652]: Dont we still want to keep the function doc for these?

File: test/test_admin.py [hunk 2]
```diff
@@ -385,15 +485,6 @@ def test_virtual_graphs(admin, music_options):
         vg.delete()
 
 
-def count_records(bd_name, conn_string):
```
Comment (buzztroll) [inline test/test_admin.py:388]: why do we want to get rid of this one?

**Review by buzztroll — CHANGES_REQUESTED** [review]
Looks great!  A couple of questions.

File: utils/run_tests.sh [hunk 3]
```diff
@@ -3,29 +3,37 @@
...
     HOST=${1}
-    PORT=${2}
+    STANDBY=${2}
+    PORT=${3}
     # Wait for stardog to be running
-    RC=1
     COUNT=0
     set +e
-    while [[ ${RC} -ne 0 ]];
+    while :
```
Comment (rmyers) [inline utils/run_tests.sh:11]: this is strange syntax, I guess it is legal, but I hate it anyway cause it makes me google things. What if instead you change to 
```
not_all_nodes_found=true
while $not_all_nodes_found
do 
         # all nodes found yay!
         not_all_nodes_found=false
```

**Review by rmyers — APPROVED** [review]
Looks good, just a few nits

**Review by buzztroll — COMMENTED** [review]
lgtm

File: stardog/admin.py (hunk 1 above)
Comment (scardena) [reply stardog/admin.py:652]: The docs were not removed btw, they were just modified a bit.

File: .circleci/config.yml [hunk 4]
```diff
@@ -19,10 +19,11 @@ jobs:
...
           name: Start containers and run tests
-          no_output_timeout: 10m
+          no_output_timeout: 15m
           command: |
             set -x
             docker-compose up --exit-code-from tests
 
 
 
+
```
[the remark this answers is not in the document]
Comment (scardena) [reply .circleci/config.yml]: hehe sorry i added a bunch of stuff for debugging here and prob a an extra space sneaked in, fixed it

File: test/test_admin.py [hunk 5]
```diff
@@ -72,14 +171,14 @@ def test_backup_all(admin):
...
+        ["sshpass", "-p", "stardogpw", "ssh", "-o", "StrictHostKeyChecking=no", "ssh://stardog@pystardog_stardog_1:2222", "--",
          "ls", "-la", "/var/opt/stardog/"],
         stdout=subprocess.PIPE, universal_newlines=True)
-    admin.backup_all(location='/tmp')
     assert '.backup' in default_backup.stdout
 
+    admin.backup_all(location='/tmp')
     tmp_backup = subprocess.run(
-        ["sshpass", "-p", "stardogpw", "ssh", "-o", "StrictHostKeyChecking=no", "ssh://stardog@pystardog_stardog:2222", "--",
+        ["sshpass", "-p", "stardogpw", "ssh", "-o", "StrictHostKeyChecking=no", "ssh://stardog@pystardog_stardog_1:2222", "--",
```
[the remark this answers is not in the document]
Comment (scardena) [reply test/test_admin.py]: I agree and addressed this, as these variables have nothing to do with the tests, but currently some tests might not work on external environments other than docker-compose, as the tests require ssh access into specific nodes for some tests. this is not always feasible to achieve, although they can easily be disabled with pytest -k 'not test'.

File: utils/run_tests.sh (hunk 3 above)
Comment (scardena) [reply utils/run_tests.sh]: Although I don't consider `while :` a strange syntax, I did address this somehow, but note that changing the expression for `$not_all_nodes_found` (or similar) might not be correct. in this case we are waiting for 2 things:
main cluster to be up and standby node to be up. I modified it a bit so it's more clear.
I am also breaking up the loop when everything is ready:
` if [[ $number_of_nodes -eq 2 && $RC -eq 0 ]]; then break; fi`
so I don't think it's necessary to change the value for the condition.

File: test/conftest.py [hunk 6]
```diff
@@ -36,14 +40,22 @@ def bulkload_content():
...
         'port': 5820,
         'username': 'admin',
         'password': 'admin'
     }
     return target_info
 
+@pytest.fixture
+def cluster_standby_node_conn_string():
+    standby_conn_string= {
+        'endpoint': 'http://' + STARDOG_HOSTNAME_STANDBY + ':5820',
```
Comment (rmyers) [inline test/conftest.py:53]: nit `f'http://{STARDOG_HOSTNAME_STANDBY}:5820

File: test/test_admin.py [hunk 7]
```diff
@@ -12,6 +12,57 @@
...
+SSH_PASS = os.environ['SSH_PASS']
+STARDOG_HOSTNAME_NODE_1 = os.environ['STARDOG_HOSTNAME_NODE_1']
+STARDOG_HOSTNAME_STANDBY = os.environ['STARDOG_HOSTNAME_STANDBY']
+
+##################
+# Help functions #
+##################
+def get_node_ip(node_hostname):
+    node_ip = subprocess.run(
+        ["sshpass", "-p", SSH_PASS, "ssh", "-o", "StrictHostKeyChecking=no", "ssh://" + SSH_USER + "@" + node_hostname + ":2222", "--",
```
Comment (rmyers) [inline test/test_admin.py:25]: nit `f"ssh://{SSH_USER}@{node_hostname}:2222"`

**Review by rmyers — APPROVED** [review]
Just a few nits, as a general rule I'd stay away from adding strings together, while it is technically legal it really shouldn't be.

**Pull request merged**